src/sha2.rs

```diff
@@ -9,10 +9,647 @@
+        ((x >> 17) | (x << 15)) ^ ((x >> 19) | (x << 13)) ^ (x >> 10)
+    }
+
+    let mut a = state[0];
+    let mut b = state[1];
+    let mut c = state[2];
+    let mut d = state[3];
+    let mut e = state[4];
+    let mut f = state[5];
+    let mut g = state[6];
+    let mut h = state[7];
+
+    let mut w = [0u32; 64];
+
+    // Sha-512 and Sha-256 use basically the same calculations which are implemented
+    // by these macros. Inlining the calculations seems to result in better generated code.
+    macro_rules! schedule {
+        ($t:expr) => (
+            w[$t] = sigma1(w[$t - 2]) + w[$t - 7] + sigma0(w[$t - 15]) + w[$t - 16];
+        )
+    }
+
+    macro_rules! round {
+        ($A:ident, $B:ident, $C:ident, $D:ident,
+         $E:ident, $F:ident, $G:ident, $H:ident, $K:ident, $t:expr) => {
+            {
+                $H += sum1($E) + ch($E, $F, $G) + $K[$t] + w[$t];
+                $D += $H;
+                $H += sum0($A) + maj($A, $B, $C);
+            }
+        }
+    }
+
+    for t in 0..16 {
+        w[t] = block[t].to_be();
+    }
+
+    // Putting the message schedule inside the same loop as the round calculations allows for
+    // the compiler to generate better code.
+    for t in range_step(0, 48, 8) {
+        schedule!(t + 16);
+        schedule!(t + 17);
+        schedule!(t + 18);
+        schedule!(t + 19);
+        schedule!(t + 20);
+        schedule!(t + 21);
+        schedule!(t + 22);
+        schedule!(t + 23);
+
+        round!(a, b, c, d, e, f, g, h, K32, t);
+        round!(h, a, b, c, d, e, f, g, K32, t + 1);
+        round!(g, h, a, b, c, d, e, f, K32, t + 2);
+        round!(f, g, h, a, b, c, d, e, K32, t + 3);
+        round!(e, f, g, h, a, b, c, d, K32, t + 4);
+        round!(d, e, f, g, h, a, b, c, K32, t + 5);
+        round!(c, d, e, f, g, h, a, b, K32, t + 6);
+        round!(b, c, d, e, f, g, h, a, K32, t + 7);
+    }
+
+    for t in range_step(48, 64, 8) {
+        round!(a, b, c, d, e, f, g, h, K32, t);
+        round!(h, a, b, c, d, e, f, g, K32, t + 1);
+        round!(g, h, a, b, c, d, e, f, K32, t + 2);
+        round!(f, g, h, a, b, c, d, e, K32, t + 3);
+        round!(e, f, g, h, a, b, c, d, K32, t + 4);
+        round!(d, e, f, g, h, a, b, c, K32, t + 5);
+        round!(c, d, e, f, g, h, a, b, K32, t + 6);
+        round!(b, c, d, e, f, g, h, a, K32, t + 7);
+    }
+
+    state[0] += a;
+    state[1] += b;
+    state[2] += c;
+    state[3] += d;
+    state[4] += e;
+    state[5] += f;
+    state[6] += g;
+    state[7] += h;
+}
+
+/// Process a block with the SHA-2 SHA-256 algorithm.
+///
+/// Chooses an implementation based on architecture,
+/// and whether or not the architecture supports SHA
+/// instruction set extensions.
+pub fn sha256_digest_block(state: &mut [u32/*; 8*/], bytes: &[u8/*; 64*/]) {
+    assert_eq!(state.len(), STATE_LEN);
+    assert_eq!(bytes.len(), BLOCK_LEN*4);
+    let (words, _): (&[u32; 16], usize) = unsafe {
+        ::std::mem::transmute(bytes)
```

Comment:
Same comment as in sha1 code regarding this resulting in unaligned memory access.